### PR TITLE
[3/N][omni, data, trainer, tests] feat: wire offline MLLM DPO inputs and OmniDirectPreferenceRayTrainer

### PR DESCRIPTION
**Pull request by wtomin**
### What does this PR do?

Builds the offline Omni DPO training path end-to-end around **adjacent chosen/rejected rows**:

- Rework `OfflineMLLMDPODataset` / `offline_mllm_dpo_collate_fn` so preference pairs stay as separate branches in `__getitem__`, then expand to adjacent rows at collate time (right-padding or `no_padding` / jagged nested tensors).
- Change `OmniDPOLoss` to consume token-level `log_probs` + batch `labels` / `ref_log_prob`, mask to sequence log-probs (optional `average_log_prob`), then split even/odd rows into chosen/rejected.
- Add `OmniDirectPreferenceRayTrainer` for offline DPO (ref inference, actor update, modality-grouped validation metrics).
- Add `GroupedMetricMean` for overall + per-modality validation aggregation.
- Keep raw multimodal token ids in `process_qwen3_omni_sample` for correct assistant labels.
- Allow explicit `train_sampler` / `val_sampler` via `create_rl_sampler(..., sampler_config=...)`.

### Test

```bash
pytest -q \
  tests/trainer/omni/test_omni_algos_on_cpu.py \
  tests/trainer/omni/test_omni_dpo_trainer_on_cpu.py \
  tests/trainer/test_metrics_utils_on_cpu.py \
  tests/utils/test_offline_mllm_dpo_dataset_on_cpu.py

### CODE REVIEW

File: verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py [hunk 1]
```diff
@@ -125,3 +129,306 @@ def configure_tokenizer(cls, model_path: str, model_config) -> Any:
...
+        Args:
+            model_config: ``OmniModelConfig`` (or compatible object with
+                ``architecture`` and ``model_stage`` for registry lookup).
+            micro_batch: ``TensorDict`` produced by the dataloader/collate path.
+            dtype: Optional parameter dtype for ``pixel_values``,
+                ``pixel_values_videos``, and ``input_features``.
+
+        Returns:
+            dict[str, Any]: Keyword arguments ready for the thinker
+            ``forward()`` call.
```
Comment (zhtmike) [inline verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py]: keep it easy to read and follow repo style

```
what is it doing

args:
    xxxx

returns:
    xxxx

somthing we need to take care
```

File: verl_omni/pipelines/model_base.py [hunk 2]
```diff
@@ -601,6 +601,58 @@ def configure_tokenizer(cls, model_path: str, model_config) -> Any:
...
+        cls,
+        model_config,
+        micro_batch: TensorDict,
+        *,
+        dtype: Optional[torch.dtype] = None,
+    ) -> dict[str, Any]:
+        """Extract standard AR/MLLM training keys from a micro-batch.
+
+        Model-family adapters override this to apply architecture-specific
+        normalization before ``model(**model_inputs)``.
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: as above

File: verl_omni/pipelines/utils.py [hunk 3]
```diff
@@ -24,23 +24,45 @@
...
+    "prepare_omni_model_inputs",
+    "prepare_omni_preference_inputs",
+    "compute_omni_preference_logps",
     "prepare_noisy_latents",
     "sample_noise_and_timesteps",
     "set_timesteps",
 ]
 
 
+def _resolve_model_family(model_config) -> str:
```
Comment (zhtmike) [inline verl_omni/pipelines/utils.py]: choose a better and meaningful name

File: verl_omni/pipelines/model_base.py [hunk 4]
```diff
@@ -601,6 +601,22 @@ def configure_tokenizer(cls, model_path: str, model_config) -> Any:
...
+        cls,
+        model_config,
+        micro_batch: TensorDict,
+        *,
+        dtype: Optional[torch.dtype] = None,
+    ) -> dict[str, Any]:
+        """Build architecture-specific forward kwargs from a micro-batch.
+
+        Each omni model family extracts the keys it needs from ``micro_batch``
+        and applies any normalization before ``model(**model_inputs)``.
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: should we mention when it is necessary, like gspo does not need this

Comment (wtomin) [reply verl_omni/pipelines/model_base.py]: I see your point. It's likely that the batch data returned by the multimodal dpo dataloader is not exactly what is needed by `model`.

I will check the batch data in details, and see if I can get rid of the `prepare_model_inputs` here.

Comment (wtomin) [reply verl_omni/pipelines/model_base.py]: @zhtmike  Refactor is done. I made small alignments to `offline_mllm_dpo_dataset` so that `prepare_model_inputs` is no longer needed.

File: verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py [hunk 5]
```diff
@@ -61,6 +61,11 @@ def configure_model(cls, module, model_config):
         module.get_input_embeddings = module.thinker.get_input_embeddings
         module.set_input_embeddings = module.thinker.set_input_embeddings
         module._no_split_modules = ["Qwen3OmniMoeThinkerTextDecoderLayer"]
+        thinker_config = getattr(module.thinker, "config", None)
+        talker_config = getattr(getattr(module, "config", None), "talker_config", None)
+        vision_start_token_id = getattr(talker_config, "vision_start_token_id", None)
+        if thinker_config is not None and vision_start_token_id is not None:
+            thinker_config.vision_start_token_id = vision_start_token_id
```
Comment (zhtmike) [inline verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py]: I think we can directly access them

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py:131]: why change BaseRayDiffusionTrainer?

File: verl_omni/utils/metrics_utils.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class GroupedMetricMean:
+    """Aggregate metric means overall and optionally by one attribute.
+
+    Each update consumes metrics that are already averaged over a batch and a
+    weight equal to the number of logical samples represented by that batch.
+    When ``group_attribute`` is set, callers pass the group value through
+    ``attributes`` and the aggregator emits both overall and per-group means.
+    When ``group_attribute`` is ``None``, only the overall means are emitted.
```
Comment (zhtmike) [inline verl_omni/utils/metrics_utils.py]: pls change to nice API format

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py
Comment (wtomin) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py:131]: Exposing `val_sampler` argument to `BaseRayDiffusionTrainer._create_dataloader`, as for some datasets, validation sampler and training sampler are both required.

File: verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py (hunk 5 above)
Comment (wtomin) [reply verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py]: I edit this because the `get_rope_index` function was called in `model.forward` in `transformers\models\qwen3_omni_moe\modeling_qwen3_omni_moe.py`:
```
def get_rope_index(...):
    spatial_merge_size = self.spatial_merge_size
    image_token_id = self.config.image_token_id
    video_token_id = self.config.video_token_id
    audio_token_id = self.config.audio_token_id
    vision_start_token_id = self.config.vision_start_token_id
    audio_start_token_id = self.config.audio_start_token_id
    position_id_per_seconds = self.config.position_id_per_seconds
```
It tries to get `vision_start_token_id` from `self.config`, and somehow `self.config` here is the thinker config. 

I can double check if this edit is needed.

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py
Comment (zhtmike) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py:131]: but you are doing omni model?

Comment (wtomin) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py:131]: Well, I let `OmniDirectPreferenceRayTrainer` inherits `DirectPreferenceRayTrainer`, and `DirectPreferenceRayTrainer` inherits `BaseRayDiffusionTrainer`.
I will disentangle them.

File: verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py (hunk 5 above)
Comment (wtomin) [reply verl_omni/pipelines/qwen3_omni/thinker_training_adapter.py]: I confirmed that this edit is not needed. The change has been reverted.

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py
Comment (wtomin) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py:131]: Disentangled

File: verl_omni/utils/metrics_utils.py (hunk 6 above)
Comment (wtomin) [reply verl_omni/utils/metrics_utils.py]: Updated

**Review by zhtmike — APPROVED** [review]
